encoding/wave/decode.go

```diff
@@ -0,0 +1,217 @@
+		if err != nil {
+			return err
+		}
+		if n != typesize {
+			return fmt.Errorf("corrupted audio")
+		}
+
+		sample := int16(binary.LittleEndian.Uint16(buf[:]))
+		*dataptr = append(*dataptr, sample)
+		bytesRead += typesize
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeFloat32(v reflect.Value, datasz uint32) error {
+	const maxval float32 = 1.0
+	const minval float32 = -1.0
+	const typesz = 4
+
+	dataptr := v.Interface().(*[]float32)
+	var bytesRead uint32
+	for bytesRead < datasz {
+		var sample float32
+		err := binary.Read(d.input, binary.LittleEndian, &sample)
+		if err != nil {
+			return fmt.Errorf("decoding floats: bytes read[%d], total[%d]: %s",
+				bytesRead,
+				datasz,
+				err)
+		}
+
+		if sample < minval || sample > maxval {
+			return fmt.Errorf(
+				"sample[%f] is outside the valid value range for a PCM float",
+				sample,
+			)
+		}
+
+		*dataptr = append(*dataptr, sample)
+		bytesRead += typesz
+	}
+
+	return nil
+}
+
+func (d *Decoder) parseRIFFHdr() (RiffHeader, error) {
+	var hdr RiffHeader
+	err := binary.Read(d.input, binary.LittleEndian, &hdr)
+	if err != nil {
+		return RiffHeader{}, fmt.Errorf("parsing riff: %s", err)
+	}
+	if string(hdr.Ident[:]) != "RIFF" {
+		return RiffHeader{}, fmt.Errorf("Invalid RIFF ident: %s", string(hdr.Ident[:]))
+	}
+	return hdr, nil
+}
+
+func (d *Decoder) parseHeader() (Header, error) {
+	riffhdr, err := d.parseRIFFHdr()
+	if err != nil {
+		return Header{}, err
+	}
+
+	// FMT chunk
+	var chunk [4]byte
+	var chunkFmt RiffChunkFmt
+
+	err = binary.Read(d.input, binary.LittleEndian, &chunk)
+	if err != nil {
+		return Header{}, fmt.Errorf("parsing fmt chunk: %s", err)
+	}
+
+	if string(chunk[:]) != "fmt " {
+		return Header{}, fmt.Errorf("Unexpected chunk type: %s", string(chunk[:]))
+	}
+
+	err = binary.Read(d.input, binary.LittleEndian, &chunkFmt)
+	if err != nil {
+		return Header{}, fmt.Errorf("parsing fmt chunk: %s", err)
+	}
+
+	if !isValidWavFormat(chunkFmt.AudioFormat) {
+		return Header{}, fmt.Errorf("Isn't an audio format: format[%d]", chunkFmt.AudioFormat)
+	}
+
+	if chunkFmt.LengthOfHeader != 16 {
+		var extraparams uint16
+		// Get extra params size
+		if err = binary.Read(d.input, binary.LittleEndian, &extraparams); err != nil {
+			return Header{}, fmt.Errorf("error getting extra fmt params: %s", err)
+		}
+		// TODO: Skip for now
+		_, err = d.input.Read(make([]byte, int(extraparams)))
+		if err != nil {
+			return Header{}, fmt.Errorf("error skipping extra params: %s", err)
+		}
+	}
+
+	var chunkSize uint32
+	for string(chunk[:]) != "data" {
+		// Read chunkID
+		err = binary.Read(d.input, binary.BigEndian, &chunk)
+		if err != nil {
+			return Header{}, fmt.Errorf("Expected data chunkid: %s", err)
+		}
+
+		err = binary.Read(d.input, binary.LittleEndian, &chunkSize)
+		if err != nil {
+			return Header{}, fmt.Errorf("Expected data chunkSize: %s", err)
+		}
+
+		// ignores LIST chunkIDs (unused for now)
+		if string(chunk[:]) != "data" {
```

Comment:
Repeating the same validation that is done on the for here seems a little odd

Reply:
at this point, I do not know if chunk is "data". Do you know a better way to handle that?

Reply:
the only one that comes to my mind is a for{} since you already validate chunk inside the loop and have other exits points to errors, but not sure if it is better.